### PR TITLE
HPCC-25262 Metrics milestone 1

### PR DESCRIPTION
**Pull request by kenrowland**
Implement and collect a counter metric

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jmetrics.hpp [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    protected:
+        std::string name;
+        std::string type;
+        MetricsReporter *pReporter = nullptr;
+};
+
+extern "C" { typedef hpccMetrics::MetricSink* (*getSinkInstance)(const char *, const IPropertyTree *pSettingsTree); }
+
+struct SinkInfo
```
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:252]: better as an opaque type define within the cpp file.

File: system/jlib/jmetrics.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::queryMetricsForReport(const std::string &sinkName)
+{
+    std::vector<std::shared_ptr<IMetric>> reportMetrics;
+
+    auto it = sinks.find(sinkName);
+    if (it != sinks.end())
```
Comment (ghalliday) [inline system/jlib/jmetrics.cpp:113]: else throw an internal error.
Also - with this implementation the name can't be optional.

File: system/jlib/jmetrics.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        //
+        // Lock the list of metrics while it's in use
+        std::unique_lock<std::mutex> lock(metricVectorMutex);   // no one else can mess with it for a bit
```
Comment (ghalliday) [inline system/jlib/jmetrics.cpp:117]: worth reserving the space in the reportMetrics vector before appending, otherwise this could be O(N^2) for very large numbers of metrics.

File: system/jlib/jmetrics.cpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+{
+    bool rc = true;
+    for (pSinkIt->first(); pSinkIt->isValid() && rc; pSinkIt->next())
+    {
+        SinkInfo *pSinkInfo;
+        Owned<IPropertyTree> pSinkTree = &pSinkIt->get(); // pSinkIt->query();
+
+        StringBuffer cfgSinkType, cfgSinkName;
+        pSinkTree->getProp("@type", cfgSinkType);  // this one is required
+        pSinkTree->getProp("@name", cfgSinkName);
```
Comment (ghalliday) [inline system/jlib/jmetrics.cpp:152]: report an error if type (or name) is missing.

File: system/jlib/jmetrics.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+bool MetricsReporter::initializeSinks(IPropertyTreeIterator *pSinkIt)
+{
+    bool rc = true;
+    for (pSinkIt->first(); pSinkIt->isValid() && rc; pSinkIt->next())
+    {
+        SinkInfo *pSinkInfo;
+        Owned<IPropertyTree> pSinkTree = &pSinkIt->get(); // pSinkIt->query();
+
+        StringBuffer cfgSinkType, cfgSinkName;
+        pSinkTree->getProp("@type", cfgSinkType);  // this one is required
```
Comment (ghalliday) [inline system/jlib/jmetrics.cpp:151]: queryProp() is generally more efficient - preferred if the IPropertyTree will definitely remain alive.

File: esp/bindings/http/platform/httpprot.cpp [hunk 6]
```diff
@@ -455,13 +455,14 @@ bool CHttpThread::onRequest()
         httpserver.setown(new CEspHttpServer(*m_socket, m_apport, m_viewConfig, getMaxRequestEntityLength()));
     }
 
-    time_t t = time(NULL);  
+    time_t t = time(NULL);
     initThreadLocal(sizeof(t), &t);
 
     m_httpserver = httpserver;
     httpserver->setSocketReturner(this);
```
Comment (ghalliday) [inline esp/bindings/http/platform/httpprot.cpp:462]: unrelated change

File: system/jlib/jmetrics.hpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+{
+    public:
+        virtual ~MetricSink() = default;
+        virtual void startCollection(MetricsReporter *pReporter) = 0;
+        virtual void stopCollection() = 0;
+        const std::string &queryName() const { return name; }
+        const std::string &queryType() const { return type; }
+
+    protected:
+        MetricSink(std::string _name, std::string _type) :
```
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:240]: not sure why std::string parameters are used here.  (The base classes don't have them.)  More normal is to pass const char *

**Review by ghalliday — CHANGES_REQUESTED** [review]
@kenrowland 
I have reviewed the code.  I commented on most things, but some items are much more significant than others.

I have a few overall comments on the current structure:

* process for reporting metrics:

Currently the process for gathering the metrics is as follows:

    auto reportMetrics = pReporter->queryMetricsForReport(name);
    MeasurementVector measurements;
    for (auto &pMetric: reportMetrics)
        pMetric->collect(measurements);
    writeMeasurementsToFile(measurements);

i.e. a vector of all the metrics is created, and then a vector of measurements.  With large numbers of metrics that will add a significant overhead.

A simpler and more efficient approach would be to directly walk the measurements and report the values by using members of the metric class e.g.

for (auto & cur : measurements)
   report(cur.queryName(), cur.queryValue(), cur....)

which avoids creating the temporary vector of measurements.

Note:
You can take this further and also avoid creating the reportMetrics list by providing a callback interface to report the metrics through.  (This is the approach of my demo branch.)  The disadvantage is that the metrics would need to be locked while the metrics are reported (unlikely to be an issue in practice as long unless the sink can block).

* Simplify types

Unless there is an example where it is useful we should assume all metrics have 8 byte unsigned values.  It would simplify the interface to metrics, no need for template classes and simplify reporting them.

* What objects types are registered.  (Not proposing changing)

I am still in two minds whether metrics should be registered with the metric reporter, or whether it should be a more general metric reporter.  Registering metrics makes it possible to filter metrics more efficiently (if that is a significant requirement), and has fewer complications with object lifetimes.  Registering reporters would reduce the overhead if large numbers of metrics are reported, and may make it easier to integrate stats.   I don't propose we change anything, but there might come a time where we want to revisit this.

File: devdoc/Metrics.rst [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+This section discusses configuration. Since Helm charts are capable of combining configuration data
+at a global level into a component's specific configuration, The combined configuration takes the
+form as shown below. Note that as the design progresses it is expected that there will be additions.
+
+::
+
+  component:
+    metrics:
+      sinks:
+        sinK:
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply devdoc/Metrics.rst:293]: This is what I had to do, based on information from Tony, in order to parse the configuration independent of the source, either environment.xml or YML (through my test program).

[the remark this answers is not in the document]
Comment (afishbeck) [reply devdoc/Metrics.rst:293]: I think my comment was about YAML that matched an existing xml based PTree.  That shouldn't be reflected in the format of the helm values themselves.  But as Gavin says that should become clear when the actual helm integration is done.

**Review by kenrowland — COMMENTED** [review]
@ghalliday Please look at these changes

**Review by ghalliday — COMMENTED** [review]
@kenrowland that's looking good.  A few (hopefully) final questions/comments.

**Comment by ghalliday** [comment]
Also, I think you misunderstood my comment about 
add_definitions( -D_METRICS_FILESINK_LIB_EXPORTS)
You do need code as you had - my comment was it should be something simpler like:
add_definitions( -DFILESINK_EXPORTS)

File: system/jlib/jmetrics.cpp (hunk 3 above)
Comment (kenrowland) [reply system/jlib/jmetrics.cpp:106]: Here is my thinking. The mutex is protecting access to the map of metrics managed by the reporter. The map contains weak pointers, so it does not maintain a reference count. When a sink needs its list of metrics to report, each metric weak pointer is converted to a shared pointer which is an atomic operation. The shared pointer is placed in a vector and returned to the sink, thus ensuring it has a reference until the sink destroys the vector. During that time if another sink requests its list, the pointer is already locked, so the conversion to shared increments the reference count. 

However, during a report let's say the component destroys the metric by clearing the shared pointer. The metric object will remain alive since the report vector contains a shared pointer to the metric and thus a reference. When the report is completed and the sink destroys the report vector, the ref count goes to 0 and the metric is destroyed (note that the weak pointer remains in the managed list by the reporter). The next time a metric report list is requested, the conversion to a shared pointer will fail and the metric is removed from the list. 

If I am missing something, please let me know.

File: system/jlib/jmetrics.cpp [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    auto it = metrics.find(pMetric->queryName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->queryName(), pMetric});
+    }
+    // todo - need to decide how to handle duplicate named metrics
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply system/jlib/jmetrics.cpp]: Clearing the shared pointer will work (see comments later). Since that will be the most efficient method, I will remove this call.

**Review by kenrowland — COMMENTED** [review]
@ghalliday Please see latest changes and comments.

File: system/jlib/jmetrics.cpp (hunk 3 above)
Comment (ghalliday) [reply system/jlib/jmetrics.cpp:106]: I agree with your overall statement.

My comment was in reference to the statement "A planned optimization is to only build this vector when there is a change".  I assumed that mean you were thinking of preserving the vector from call to call.  If that was done the vector would contain shared_ptrs, which would mean the weak_ptr never returned null.

File: system/jlib/jmetrics.hpp [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    std::vector<std::shared_ptr<IMetric>> queryMetricsForReport(const std::string &sinkName);
+
+protected:
+    void initializeSinks(IPropertyTreeIterator *pSinkIt);
+    static MetricSink *getSinkFromLib(const char *type, const char *sinkName, const IPropertyTree *pSettingsTree);
+
+protected:
+    StringBuffer componentPrefix;
+    StringBuffer globalPrefix;
+    std::map<std::string, std::shared_ptr<SinkInfo>> sinks;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply system/jlib/jmetrics.hpp:186]: I think use std::unique_ptr until there is a reason to make it shared.

File: system/jlib/jmetrics.hpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    }
+};
+
+
+
+class jlib_decl MetricSink
+{
+public:
+    virtual ~MetricSink() = default;
+    virtual void startCollection(MetricsReporter *pReporter) = 0;
```
Comment (ghalliday) [inline system/jlib/jmetrics.hpp:148]: minor: This looks like it should be passed to the getSinkInstance() function, but I wouldn't change for this PR.

File: system/metrics/sinks/file/fileSink.hpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+#include "jstring.hpp"
+#include "jfile.ipp"
+#include "jsem.hpp"
+#include <thread>
+#include <condition_variable>
+
+
+using namespace hpccMetrics;
+
+class DECL_EXPORT FileMetricSink : public MetricSink
```
Comment (ghalliday) [inline system/metrics/sinks/file/fileSink.hpp:28]: This should be based on a conditional define.  You had code previously to implement it, but I commented that the name was verbose/did not match the general pattern used elsewhere.

**Review by ghalliday — COMMENTED** [review]
@kenrowland lloking very good.  A few final minor comments.

**Review by kenrowland — COMMENTED** [review]
@ghalliday Please review latest changes. Thanks.

**Review by ghalliday — APPROVED** [review]
@kenrowland great.  Please squash that last commit and I will merge.
